mne/beamformer/_ap.py

```diff
@@ -0,0 +1,712 @@
+    else:
+        idx, oris, poss, gain_active, gain_dip, sol = \
+            _fixed_ori_ap(wh_data, gain, nsources, forward,
+                          max_iter=max_iter)
+        X = sol
+        dip_ind = idx
+
+    gain_active = whitener @ gain_active.reshape(gain.shape[0], -1)
+    explained_data = gain_dip @ sol
+    m_estimate = whitener @ explained_data
+    var_exp = _log_exp_var(wh_data, m_estimate)
+    tstep = np.median(np.diff(times)) if len(times) > 1 else 1.
+    dipoles = _make_dipoles_sparse(
+        X, idx, forward, times[0], tstep, wh_data,
+        gain_active, active_is_idx=True)
+    for dipole, ori in zip(dipoles, oris):
+        signs = np.sign((dipole.ori * ori).sum(-1, keepdims=True))
+        dipole.ori *= signs
+        dipole.amplitude *= signs[:, 0]
+    logger.info('[done]')
+
+    return dipoles, explained_data, var_exp, dip_ind, oris, poss
+
+
+def _make_explained_evoke(evoked, picks, explained_data_mat, residual=False):
+    """Create a new Evoked object containing explained/residual data."""
+    n_evoked = evoked.copy()
+    n_evoked = n_evoked.pick(picks)
+    n_evoked.data = (
+        n_evoked.data - explained_data_mat if residual else explained_data_mat)
+    active_projs = [proj for proj in n_evoked.info['projs'] if proj['active']]
+    for proj in active_projs:
+        proj['active'] = False
+    n_evoked.add_proj(active_projs, remove_existing=True)
+    n_evoked.apply_proj()
+    return n_evoked
+
+
+@verbose
+def alternating_projections(evoked, forward, nsources, noise_cov=None,
+                            max_iter=6, return_residual=True,
+                            return_active_info=False, verbose=None,
+                            force_no_rep=False):
+    """Alternating Projections sources localization method.
+
+    Compute Alternating Projections (AP) on evoked data.
+
+    The implementation follows :footcite:t:`AdlerEtAl2019,AdlerEtAl2022`.
+
+    Parameters
+    ----------
+    evoked : instance of Evoked
+        Evoked object containing data to be localized.
+    forward : instance of Forward
+        Forward operator.
+    nsources : int
+        The number of dipoles to estimate.
+    noise_cov : instance of Covariance, optional
+        The noise covariance matrix, used for whitening the evoked signal.
+        The default is None. If None is given, the noise covariance matrix
+        is created ad-hoc (See: mne.cov.make_ad_hoc_cov).
+    max_iter : int, optional
+        Maximal iteration over estimated dipoles. The default is 6.
+    return_residual : bool, optional
+        If True, appends residual, explained_data and var_exp to output.
+        The default is True.
+    return_active_info : bool, optional
+        If True, appends estimated source's information
+        (indices,coordinates,orientation). The default is False.
+    %(verbose)s
+    force_no_rep : bool, optional
+            Forces no repetition of estinated dipoles.
+
+    Returns
+    -------
```

Comment:
see how the Returns section of the rap_music function is structured. Output is never a list (maybe a tuple). Now See rap_music how we suggest to do it.